_posts/2021-05-30-spring-elasticache-redis.md

```diff
@@ -0,0 +1,230 @@
+
+```yaml
+Parameters:
+
+  InstanceType:
+    Description: Instance Type to use for Cache Cluster
+    Type: String
+    Default: cache.t2.micro
+
+
+Resources:
+
+  ProductCache:
+    Type: AWS::ElastiCache::CacheCluster
+    Properties:
+      ClusterName: product-cache
+      Engine: "Redis"
+      EngineVersion: "6.x"
+      CacheNodeType: !Ref InstanceType
+      NumCacheNodes: 1
+      VpcSecurityGroupIds: ["sg-xxxx"]
+
+  UserCache:
+    Type: AWS::ElastiCache::CacheCluster
+    Properties:
+      ClusterName: user-cache
+      Engine: "Redis"
+      EngineVersion: "6.x"
+      CacheNodeType: !Ref InstanceType
+      NumCacheNodes: 1
+      VpcSecurityGroupIds: [ "sg-xxxx" ]
+```
+
+Then instead of giving cluster names, we only need to provide the stack name. Say the stack name
+is `example-stack`:
+
+```yaml
+cloud:
+  aws:
+    stack:
+      name: example-stack
+```
+
+Spring Cloud AWS retrieves all the cache clusters from our stack and builds `CacheManager` with
+the names of the resources as cache names instead of cluster names. With this, we need to
+update our cache names in our application to resource names instead of cluster names:
+
+```java
+@CacheConfig(cacheNames = "ProductCache")
+public class ProductService {
+    //...
+}
+
+@CacheConfig(cacheNames = "UserCache")
+public class UserService {
+    //...
+}
+```
+
+Make sure to add the following dependency when using the stack name approach:
+
+```groovy
+implementation 'com.amazonaws:aws-java-sdk-cloudformation'
+```
+
+
+## How Does Spring Cloud AWS Configure the `CacheManager`?
+
+In this section, we will dive a bit deeper into the inner workings of the Spring Cloud AWS
+and see how it autoconfigures the cache for us.
+
+As we know that for caching to work in a Spring application we need a `CacheManager` bean. The job of Spring Cloud AWS is to essentially create that bean for us.
+
+Let's look at the steps it performs along with the classes involved in building `CacheManager`:
+
+![Spring Cloud AWS](/assets/img/posts/spring-elasticache-redis/spring-cloud-classes.png)
+
+* When our application starts in the AWS environment, `ElastiCacheAutoConfiguration` reads cluster names from the
+  `application.yml` and passes them on to the `ElastiCacheCacheConfigurer` object.
+* Then `ElastiCacheCacheConfigurer` creates the `CacheManager` with the help of `ElastiCacheFactoryBean` class.
+* `ElastiCacheFactoryBean` scans the ElastiCache in the same availability zone and retrieves the host and port
+  names of the nodes. To allow our service to scan ElatiCache we need to provide `AmazonElastiCacheReadOnlyAccess` permission
+  to our service or `AWSCloudFormationReadOnlyAccess` if we are using the stack name approach.
+* Also `ElastiCacheFactoryBean` passes this host and port to `RedisCacheFactory` which then uses Redis clients such as Lettuce to create the connection object which then actually establishes with the nodes and perform required operations.
+
```

Comment:
Please relook at the article in totality. At the end of the article the reader should be able to run the application with the cache configured. 
Somewhere I expect to see caching configuration using Java configuration along with creation of a Spring boot project.

Reply:
Sure, will do that. I have omitted one configuration which is important for running the application will add that too.

For configuring cache apart from auto configuration the only other approach available with Spring Cloud is the annotations based but they have deprecated that. If I have to showcase Java configuration that would mean going back to the non spring cloud approach where we configure caching factories and cache managers by ourselves. Please let me know if I didn't get your point. 

Also, can you please elaborate the `creation of spring boot project` point?

Reply:
Where will the reader start from? You need to mention that step.

"The caching support for Spring Cloud AWS  uses Spring Data Redis for Redis caches". So you need this configuration for running the example. Keep it simple. Just describe the steps to set up and run the example.